File: TypeScript/src/Day7.ts
import Day from "./Day";
import * as Util from "./Util";

interface BagType{
    appearance: string;
    color: string;
}

interface LuggageRule{
    containingBag: BagType;
    containedBags: Map<BagType, number>;
}

function equalBagTypes(bagType: BagType, otherBag: BagType): boolean {
    return otherBag.appearance == bagType.appearance
        && otherBag.color == bagType.color;
}

class Day7 implements Day<LuggageRule[]>{
    parseInput(text: string): LuggageRule[] {
        return text.split(/\r?\n/)
            .map((line: string) => this.parseLuggageRule(line))
            .filter(Util.notEmpty);
    }

    private parseLuggageRule(ruleText: string): LuggageRule | null{
        const pattern: RegExp = /(\w+) (\w+) bags contain(( no other bags)|(,? \d+ \w+ \w+ bags?)+)/;
        const match = ruleText.match(pattern);

        if(!match){
            return null;
        }

        const containingBag: BagType = {appearance: match[1], color: match[2]}
        const containedBags: Map<BagType,number> = this.parseContainedBags(match[3]);
        return {containingBag: containingBag, containedBags: containedBags};
    }

    private parseContainedBags(containedBagsText: string): Map<BagType,number>{
        let containedBags: Map<BagType,number> = new Map<BagType,number>();

        if(containedBagsText == " no other bags"){
            return containedBags;
        }

        const containingBagDefinitions: string[] = containedBagsText.split(',');
        containingBagDefinitions
            .map((definition: string) => this.parseContainingBagSpecification(definition))
            .filter(Util.notEmpty)
            .forEach((bagSpecification: [BagType, number]) => containedBags.set(bagSpecification[0], bagSpecification[1]));

            return containedBags;
    }

    private parseContainingBagSpecification(definition: string): [BagType, number] | null{
        const pattern: RegExp = /(\d+) (\w+) (\w+) bags?/;
        const match = definition.match(pattern);

        if(!match){
            return null;
        }

        const containedBag: BagType = {appearance: match[2], color: match[3]};
        const numberOfBags: number = parseInt(match[1]);

        return [containedBag, numberOfBags];
    }

    solvePart1(input: LuggageRule[]): string {
        const myBag: BagType = {appearance: "shiny", color: "gold"};
        const containedBagsGraph: Map<string, Set<string>> = this.toContainedGraph(input);
        const containingBags: Set<string> = this.descendants(containedBagsGraph, JSON.stringify(myBag));
        return containingBags.size.toString();
    }

    //string because javascript does not allow to redefine equality for the map.
    private toContainedGraph(rules: LuggageRule[]): Map<string, Set<string>>{
        let graph = new Map<string, Set<string>>();
        for (let rule of rules){
            for (let containedBag of rule.containedBags.keys()){
                this.addDirectedEdge(graph, JSON.stringify(containedBag), JSON.stringify(rule.containingBag));
            }
        }
        return graph;
    }

    private addDirectedEdge<TVertex>(graph: Map<TVertex, Set<TVertex>>, startVertex: TVertex, endVertex: TVertex){
        if(!graph.has(startVertex)){
            graph.set(startVertex, new Set<TVertex>([endVertex]));
        }

        let currentChildren: Set<TVertex> = graph.get(startVertex) as Set<TVertex>;
        currentChildren.add(endVertex);
    }

    private descendants<TValue>(dac: Map<TValue, Set<TValue>>, vertex: TValue): Set<TValue>{
        const descendants: Set<TValue> = new Set<TValue>();

        if(!dac.has(vertex)){
            return descendants;
        }

        const children: Set<TValue> = dac.get(vertex) as Set<TValue>;

        for(let child of children){
            this.addWithDescendants(descendants, dac, child);
        }

        return descendants;
    }

    private addWithDescendants<TValue>(set: Set<TValue>, dac: Map<TValue, Set<TValue>>, vertex: TValue){
        if(set.has(vertex)){
            return;
        }
        set.add(vertex);

        if(!dac.has(vertex)){
            return;
        }

        const children: Set<TValue> = dac.get(vertex) as Set<TValue>;

        for(let child of children){
            this.addWithDescendants(set, dac, child);
        }
    }

    solvePart2(input: LuggageRule[]): string {
        const myBag: BagType = {appearance: "shiny", color: "gold"};
        const myBagKey: string = JSON.stringify(myBag);
        const allContainedBags: Map<string,Map<string,number>> = this.totalContainedBags(input);
        if(!allContainedBags.has(myBagKey)){
            return "404 Bag not Found!"
        }
        const containedBags: Map<string, number> = allContainedBags.get(myBagKey) as Map<string, number>;
        const totalBagCount: number = Array.from(containedBags.values()).reduce((sum: number, item: number) => sum + item, 0);
        return totalBagCount.toString();
    }

    private totalContainedBags(rules: LuggageRule[]): Map<string,Map<string,number>>{
        const ruleDac: Map<string,Map<string,number>> = this.toContainingGraph(rules);
        const containedBags: Map<string,Map<string,number>> = new Map<string,Map<string,number>>();
        for (let bagType of ruleDac.keys()){
            this.addTotalContainedBags(containedBags, ruleDac, bagType);
        }
        return containedBags;
    }

    private toContainingGraph(rules: LuggageRule[]): Map<string,Map<string,number>>{
        const ruleDac: Map<string,Map<string,number>> = new Map<string,Map<string,number>>();
        for(let rule of rules){
            ruleDac.set(JSON.stringify(rule.containingBag), this.toStringMap(rule.containedBags));
        }
        return ruleDac;
    }

    private toStringMap<TKey, TValue>(objectMap: Map<TKey, TValue>): Map<string, TValue>{
        const stringMap: Map<string, TValue> = new Map<string, TValue>();
        for(let element of objectMap){
            stringMap.set(JSON.stringify(element[0]), element[1]);
        }
        return stringMap;
    }

    private addTotalContainedBags(containedBags: Map<string,Map<string,number>>, ruleDac: Map<string,Map<string,number>>, bagType: string){
        if(containedBags.has(bagType) || !ruleDac.has(bagType)){
            return;
        }

        const totalBagsInBagType: Map<string, number> = new Map<string, number>();
        const bagsInBagType: Map<string, number> = ruleDac.get(bagType) as Map<string, number>;

        for (let element of bagsInBagType){
            let containedBag: string = element[0];
            let numberOfBags: number = element[1];

            if(!containedBags.has(containedBag)){
                this.addTotalContainedBags(containedBags, ruleDac, containedBag);
            }

            this.addBagTotals(totalBagsInBagType, containedBag, numberOfBags, containedBags);
        }

        containedBags.set(bagType, totalBagsInBagType);
    }

    private addBagTotals(totals: Map<string, number>, bagType: string, count: number, containedBags: Map<string,Map<string,number>>){
        Util.addCount(totals, bagType, count);
        let containedBagTotals: Map<string, number> = containedBags.get(bagType) as Map<string, number>;
        for(let element of containedBagTotals){
            let containedBag: string = element[0];
            let numberOfBags: number = element[1];
            Util.addCount(totals, containedBag, numberOfBags * count);
        }
    }
}



export default Day7;

File: Rust/adventOfCode2020/src/day/day7.rs
use std::collections::{HashMap, HashSet};
use std::hash::Hash;


#[derive(PartialEq, Eq, Hash, Debug, Clone, Default)]
struct BagType{
    appearance: String,
    color: String,
}

#[derive(PartialEq, Eq, Hash, Debug, Clone, Default)]
pub struct LuggageRule{
    containing_bag: BagType,
    contained_bags: Vec<(BagType, usize)>,
}


pub struct Day7 {}

impl super::Day for Day7{
    type PuzzleInput = Vec<LuggageRule>;

    fn parse_input(&self, text: std::string::String) -> Self::PuzzleInput {
        text.lines()
            .filter_map(|line| parsed_luggage_rule(line))
            .collect()
    }

    fn solve_part1(&self, rules: Self::PuzzleInput) -> std::string::String {
        let my_bag = BagType {appearance: String::from("shiny"), color: String::from("gold")};
        let graph = contained_graph(&rules);
        let containing_bags = descendants(&graph, &my_bag);
        let result = containing_bags.len();
        result.to_string()
    }

    fn solve_part2(&self, rules: Self::PuzzleInput) -> std::string::String {
        let my_bag = BagType {appearance: String::from("shiny"), color: String::from("gold")};
        let graph = containing_graph(&rules);
        let containing_bags = total_contained_bags(&graph, &my_bag);
        let result: usize = containing_bags.values().sum();
        result.to_string()
    }
}

fn parsed_luggage_rule(line: &str) -> Option<LuggageRule>{
    lazy_static! {
        static ref LUGGAGE_RULE_RE: regex::Regex = regex::Regex::new(r"(\w+) (\w+) bags contain(( no other bags)|(,? \d+ \w+ \w+ bags?)+)").unwrap();
    }
    let captures: regex::Captures = LUGGAGE_RULE_RE.captures(line)?;
    let containing_bag = BagType{ appearance: captures[1].to_owned(), color: captures[2].to_owned()};
    let contained_bags = parsed_contained_bags(&captures[3])?;
    Some(LuggageRule {containing_bag, contained_bags })
}

fn parsed_contained_bags(bags_text: &str) -> Option<Vec<(BagType, usize)>>{
    if bags_text == " no other bags"{
        return Some(Vec::new());
    }

    let contained_bag_definitions = bags_text.split(',');
    let contained_bag_specifications = contained_bag_definitions
        .filter_map(|bag_definition| parsed_contained_bag(bag_definition))
        .collect();
        Some(contained_bag_specifications)
}

fn parsed_contained_bag(bag_text: &str) -> Option<(BagType, usize)>{
    lazy_static! {
        static ref BAG_RE: regex::Regex = regex::Regex::new(r"(\d+) (\w+) (\w+) bags?").unwrap();
    }
    let captures: regex::Captures = BAG_RE.captures(bag_text)?;
    let contained_bag_type = BagType {appearance: captures[2].to_owned(), color: captures[3].to_owned()};
    let number_of_contained_bags = captures[1].parse::<usize>().ok()?;
    Some((contained_bag_type, number_of_contained_bags))
}

fn contained_graph(rules: &[LuggageRule]) -> HashMap<BagType, Vec<(BagType, usize)>>{
    let mut graph = HashMap::new();
    for rule in rules.iter(){
        let containing_bag = &rule.containing_bag;
        for (contained_bag, bag_count) in &rule.contained_bags{
            add_directed_edge(&mut graph, contained_bag, containing_bag, bag_count);
        }
    }
    graph
}

fn add_directed_edge<TVertex: Eq + Hash + Clone, TEdgeData: Copy>(graph: &mut HashMap<TVertex, Vec<(TVertex, TEdgeData)>>, start_vertex: &TVertex, end_vertex: &TVertex, edge_data: &TEdgeData){
    let children = graph.get_mut(start_vertex);
    match children{
        Some(ch) => ch.push((end_vertex.to_owned(), edge_data.to_owned())),
        None => {graph.insert(start_vertex.to_owned(), vec![(end_vertex.to_owned(), edge_data.to_owned())]);},
    }
}

fn descendants<'a, TVertex: Hash + Eq, TEdgeData>(graph: &'a HashMap<TVertex, Vec<(TVertex, TEdgeData)>>, node: &TVertex) -> HashSet<&'a TVertex>{
    let mut descendant_set = HashSet::new();
    add_descendants(graph, node, &mut descendant_set);
    descendant_set
}

fn add_descendants<'a, TVertex: Hash + Eq, TEdgeData>(graph: &'a HashMap<TVertex, Vec<(TVertex, TEdgeData)>>, node: &TVertex, nodes: &mut HashSet<&'a TVertex>){
    let maybe_children = graph.get(node);
    if let Some(children) = maybe_children {
        for (child, _edge_data) in children{
            nodes.insert(child);
            add_descendants(graph, child, nodes);
        }
    }
}

fn containing_graph(rules: &[LuggageRule]) -> HashMap<BagType, Vec<(BagType, usize)>>{
    let mut graph = HashMap::new();
    for rule in rules.iter(){
        let containing_bag = &rule.containing_bag;
        for (contained_bag, bag_count) in &rule.contained_bags{
            add_directed_edge(&mut graph, containing_bag, contained_bag, bag_count);
        }
    }
    graph
}

fn total_contained_bags(rules_dac: &HashMap<BagType, Vec<(BagType, usize)>>, node: &BagType) -> HashMap<BagType, usize>{
    let mut total_contained_bag_by_node: HashMap<BagType, HashMap<BagType, usize>> = HashMap::new();
    add_contained_bags(&mut total_contained_bag_by_node, node, rules_dac);
    total_contained_bag_by_node.remove(node).unwrap()
}

fn add_contained_bags(total_contained_bags: &mut HashMap<BagType, HashMap<BagType, usize>>, node: &BagType, rules_dac: &HashMap<BagType, Vec<(BagType, usize)>>){
    let maybe_children = rules_dac.get(node);
    if let Some(children) = maybe_children{
        let mut contained_bags: HashMap<BagType, usize> = HashMap::new();
        for (child, bag_count) in children{
            if !total_contained_bags.contains_key(child){
                add_contained_bags(total_contained_bags,child, rules_dac);
            }
            for (bag, &count) in total_contained_bags.get(child).unwrap(){
                add_to_bags( &mut contained_bags, bag, count * bag_count);
            }
            add_to_bags(&mut contained_bags, child, *bag_count);
        }
        total_contained_bags.insert(node.to_owned(), contained_bags);
    } else {
        total_contained_bags.insert(node.to_owned(), HashMap::new());
    }
}

fn add_to_bags(contained_bags: &mut HashMap<BagType, usize>, bag_type: &BagType, bag_count: usize){
    match contained_bags.get_mut(bag_type){
        Some(current_count) => *current_count += bag_count,
        None => {contained_bags.insert(bag_type.to_owned(), bag_count);},
    }
}



#[cfg(test)]
mod day7_tests {
    use super::*;
    use crate::input;
    use crate::day;

    #[test]
    fn correct_part1() {
        let day: Box<dyn day::DaySolver> = Box::new(Day7{});
        let problem_input = input::puzzle_input(&input::PuzzleConfiguration{day: 7, part: 1}).unwrap();
        let expected_result = String::from("161");
        let actual_result = day.solve_part1(problem_input);
        assert_eq!(actual_result, expected_result);
    }

    #[test]
    fn correct_part2() {
        let day: Box<dyn day::DaySolver> = Box::new(Day7{});
        let problem_input = input::puzzle_input(&input::PuzzleConfiguration{day: 7, part: 2}).unwrap();
        let expected_result = String::from("30899");
        let actual_result = day.solve_part2(problem_input);
        assert_eq!(actual_result, expected_result);
    }
}